### DISCUSSION TITLE
Drift in position server using external VICON position

### DISCUSSION THREAD

**Discussion by djd40k** [bitcraze/discussions#119] [Q&A]
I am using a crazyflie 2.1 with Vicon Tracker for external positioning.  The position from the cf1 object is consistent in ROS and remains steady with a standard deviation of the x, y, z position of about 3e-4 over time.  however, the CF1/pose topic shows a significant drift in the position estimate of the crazyflie.  It appears that the drift is occurring in the location server or kalman filter, but I have been unable to determine why this is happening.  When I attach a flow deck the drift is decreased, but does not go away.  The crazyflie reports to have passed its sensor checks on the consol.  What can I do to preserve the values reported from the external position for the pose estimate?

**Comment by krichardsson** [comment]
Could you tell us a bit more about your setup? Are you using Crazyswarm or the pyhon lib?

**Reply by djd40k** [reply]
I am using crazyswarm on Ubuntu 20.04 with ROS Noetic.  The crazyflie firmware is 2022.05 and the Python version is 3.7.  I am using the Crazyswarm Hover_swarm.launch and the hello_world.py Python script.

Here are contents of the files I have been editing:
CrazyflieTypes.yaml

crazyflieTypes:
  default:
    bigQuad: False
    batteryVoltageWarning: 3.8  # V
    batteryVoltateCritical: 3.7 # V
    markerConfiguration: 0 # 0 for unique marker (first setup) # change to 3 for single marker tracking
    dynamicsConfiguration: 0
    firmwareParams:
      #kalman:
        #pNAcc_xy: 0.005 # 4.0 from below # default 0.5
        #pNAcc_z: 0.001 # 8.0 from below #  default 1.0
        #pNPos: 0.001 # 
        #mNGyro_rollpitch: 0.1 # default 0.1
        #mNGyro_yaw: 0.1 # default 0.1
        #robustTdoa: 0  # default 0
        #robustTwr: 0 # default 0
      ctrlMel:
        kp_xy: 0.4 # original 0.4 # p for position
        kd_xy: 0.2 # original 0.2 # 0.35 # d for position
        ki_xy: 0.05 # original 0.05
        i_range_xy: 2.0
        kR_xy: 70000 # p for attitude
        kw_xy: 20000 # d for attitude
        kR_z: 70000 # 70000 # p for yaw
        kw_z: 30000 # 30000  # d for yaw
        ki_m_z: 500 #0 #20000     # i for yaw
        i_range_m_z: 1500 # i range for yaw
        kd_omega_rp: 230 #working is 300#  original 200 roll and pitch angular velocity d gain
        kp_z: 1.25  # original 1.25 # working at 1.25
        kd_z: 0.4 # original  0.4 # working at 0.7
        ki_z: 0.05 #original 0.05
        i_range_z: 0.4
        mass: 0.040 # original 0.032
        massThrust: 132000
  defaultSingleMarker:
    bigQuad: False
    batteryVoltageWarning: 3.8  # V
    batteryVoltateCritical: 3.7 # V
    markerConfiguration: 3 # changed to 3 for single marker tracking
    dynamicsConfiguration: 0
    firmwareParams:
      kalman:
        pNAcc_xy: 4.0 # default 0.5
        pNAcc_z: 8.0 # default 1.0
        mNGyro_rollpitch: 0.2 # default 0.1
        mNGyro_yaw: 0.1 # default 0.1
      ctrlMel:
        kp_xy: 0.4 #0.6 # p for position
        kd_xy: 0.2 # 0.35 # d for position
        ki_xy: 0.05
        i_range_xy: 2.0
        kR_xy: 70000 # p for attitude
        kw_xy: 20000 # d for attitude
        kR_z: 60000 # 70000 # p for yaw
        kw_z: 12000 # 30000  # d for yaw
        ki_m_z: 500 #0 #20000     # i for yaw
        i_range_m_z: 1500 # i range for yaw
        kd_omega_rp: 200 # roll and pitch angular velocity d gain
        kp_z: 1.25
        kd_z: 0.4
        ki_z: 0.05
        i_range_z: 0.4
        mass: 0.032
        massThrust: 132000
  CF21SingleMarker:
    bigQuad: False
    batteryVoltageWarning: 3.8  # V
    batteryVoltateCritical: 3.7 # V
    markerConfiguration: 3 # changed to 3 for single marker tracking
    dynamicsConfiguration: 0
    firmwareParams:
      kalman:
        pNAcc_xy: 1.0 # default 0.5
        pNAcc_z: 2.0 # default 1.0
        mNGyro_rollpitch: 0.1 # default 0.1
        mNGyro_yaw: 0.1 # default 0.1
      ctrlMel:
        kp_xy: 0.4 #0.6 # p for position
        kd_xy: 0.2 # 0.35 # d for position
        ki_xy: 0.05
        i_range_xy: 2.0
        kR_xy: 70000 # p for attitude
        kw_xy: 20000 # d for attitude
        kR_z: 60000 # 70000 # p for yaw
        kw_z: 12000 # 30000  # d for yaw
        ki_m_z: 500 #0 #20000     # i for yaw
        i_range_m_z: 1500 # i range for yaw
        kd_omega_rp: 200 # roll and pitch angular velocity d gain
        kp_z: 1.25
        kd_z: 0.4
        ki_z: 0.05
        i_range_z: 0.4
        mass: 0.032
        massThrust: 132000
  medium:
    bigQuad: True
    batteryVoltageWarning: 7.6  # V
    batteryVoltateCritical: 7.4 # V
    markerConfiguration: 1
    dynamicsConfiguration: 0
    firmwareParams:
      ctrlMel:
        kp_xy: 2.0 # p for position
        kd_xy: 1.0 # d for position
        ki_xy: 0.1 # i for position
        i_range_xy: 2.0 # range for position
        kR_xy: 50000 # 50000 # p for attitude
        kw_xy: 16000 # 16000 # d for attitude
        kR_z: 60000 # p for yaw
        kw_z: 12000 # d for yaw
        ki_m_z: 500 # i for yaw
        i_range_m_z: 1500 # i range for yaw
        kd_omega_rp: 200 # roll and pitch angular velocity d gain
        kp_z: 3 # 1.25
        kd_z: 1.0
        ki_z: 0.5 # 0.05
        i_range_z: 2.0
        mass: 0.110
        massThrust: 23000 # 23000 - full battery
  large:
    bigQuad: True
    batteryVoltageWarning: 11.4  # V
    batteryVoltateCritical: 11.1 # V
    markerConfiguration: 2
    dynamicsConfiguration: 0
    firmwareParams:
      ctrlMel:
        kp_xy: 8.0 # 0.4 #0.6 # p for position
        kd_xy: 3.0 # 0.2 # 0.35 # d for position
        ki_xy: 1.0 # 0.05
        i_range_xy: 2.0
        kR_xy: 30000 # 40000 # p for attitude
        kw_xy: 9000 # 13000 # d for attitude
        kR_z: 40000 # 70000 # p for yaw
        kw_z: 10000 # 30000  # d for yaw
        ki_m_z: 500 #0 #20000     # i for yaw
        i_range_m_z: 1500 # i range for yaw
        kd_omega_rp: 100 # roll and pitch angular velocity d gain
        kp_z: 8.0 # 1.25
        kd_z: 3.0 # 0.4
        ki_z: 2.0 # 0.05
        i_range_z: 4.0
        mass: 0.380
        massThrust: 5400 # 5500 - full battery
numMarkerConfigurations: 4
markerConfigurations:
  "0":  # for standard Crazyflie
    numPoints: 4
    offset: [0.0, -0.01, -0.04]
    points:
      "0": [-0.0141609,0.0333191,0.0274661]
      "1": [0.024467,-0.00525394,0.0275377]
      "2": [-0.00269392,0.00239877,0.0400294]
      "3": [0.0281347,0.0377976,0.0258931]
  "1": # medium frame
    numPoints: 4
    offset: [0.0, 0.0, -0.03]
    points:
      "0": [-0.00896228,-0.000716753,0.0716129]
      "1": [-0.0156318,0.0997402,0.0508162]
      "2": [0.0461693,-0.0881012,0.0380672]
      "3": [-0.0789959,-0.0269793,0.0461144]
  "2": # big frame
    numPoints: 4
    offset: [0.0, 0.0, -0.06]
    points:
      "0": [0.0558163,-0.00196302,0.0945539]
      "1": [-0.0113941,0.00945842,0.0984811]
      "2": [-0.0306277,0.0514879,0.0520456]
      "3": [0.0535816,-0.0400775,0.0432799]
  "3":  # for standard Crazyflie with a single marker
    numPoints: 1
    offset: [0.0, -0.01, -0.04]
    points:
      "0": [0.0177184,0.0139654,0.0557585]
numDynamicsConfigurations: 1
dynamicsConfigurations:
  "0":
    maxXVelocity: 2.0
    maxYVelocity: 2.0
    maxZVelocity: 3.0
    maxPitchRate: 20.0
    maxRollRate: 20.0
    maxYawRate: 10.0
    maxRoll: 1.4
    maxPitch: 1.4
    maxFitnessScore: 0.001

Hello_world.py
"""Takeoff-hover-land for one CF. Useful to validate hardware config."""

from pycrazyswarm import Crazyswarm


TAKEOFF_DURATION =  0.1  # good is 2.5 with height of 1.5
HOVER_DURATION = 0.0


def main():
    swarm = Crazyswarm()
    timeHelper = swarm.timeHelper
    cf = swarm.allcfs.crazyflies[0]


    cf.takeoff(targetHeight=0.4, duration=TAKEOFF_DURATION)
    timeHelper.sleep(TAKEOFF_DURATION + HOVER_DURATION)
    cf.land(targetHeight=-0.10, duration=3)
    timeHelper.sleep(TAKEOFF_DURATION)


if __name__ == "__main__":
    main()


Hover_swarm.launch
<?xml version="1.0"?>
<launch>
  <arg name="joy_dev" default="/dev/input/js0" />

  <rosparam command="load" file="$(find crazyswarm)/launch/crazyflieTypes.yaml" />
  <rosparam command="load" file="$(find crazyswarm)/launch/crazyflies.yaml" />

  <node pkg="crazyswarm" type="crazyswarm_server" name="crazyswarm_server" output="screen" >
    <rosparam>
      # Logging configuration (Use enable_logging to actually enable logging)
      genericLogTopics: ["log1"]
      genericLogTopicFrequencies: [10]
      genericLogTopic_log1_Variables: ["stateEstimate.x", "ctrltarget.x" ]
      # firmware parameters for all drones (use crazyflieTypes.yaml to set per type, or
      # allCrazyflies.yaml to set per drone)
      firmwareParams:
        commander:
          enHighLevel: 1
        stabilizer:
          estimator: 2 # 1: complementary, 2: kalman
          controller: 1 # 1: PID, 2: mellinger
        ring:
          effect: 16 # 6: double spinner, 7: solid color, 16: packetRate
          solidBlue: 255 # if set to solid color
          solidGreen: 0 # if set to solid color
          solidRed: 0 # if set to solid color
          headlightEnable: 0
        locSrv:
          extPosStdDev: 1.0e-4
          extQuatStdDev: 3.0e-3 #Previous Working 3.0e-3#  original 0.5e-1
        kalman:
          resetEstimation: 1
      # tracking
      motion_capture_type: "vicon" # one of none,vicon,optitrack,optitrack_closed_source,qualisys,vrpn
      object_tracking_type: "motionCapture" # one of motionCapture,libobjecttracker
      send_position_only: True # set to False to send position+orientation; set to True to send position only
      motion_capture_host_name: "192.168.10.22"
      # motion_capture_interface_ip: "192.168.10.22" # optional for optitrack with multiple interfaces
      save_point_clouds: "/home/david/crazyswarm/bin" # set to a valid path to log mocap point cloud binary file.
      print_latency: False
      write_csvs: True
      force_no_cache: False
      enable_parameters: True
      enable_logging: True
      enable_logging_pose: True
    </rosparam>
  </node>

  <node name="joy" pkg="joy" type="joy_node" output="screen">
    <param name="dev" value="$(arg joy_dev)" />
  </node>

  <node pkg="crazyswarm" type="crazyswarm_teleop" name="crazyswarm_teleop" output="screen">
    <param name="csv_file" value="$(find crazyswarm)/launch/figure8_smooth.csv" />
    <param name="timescale" value="0.8" />
  </node>

  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find crazyswarm)/launch/test.rviz"/>

  
  

</launch>

**Comment by knmcguire** [comment]
Hi! this seems to be an issue of Crazyswarm that we don't maintain ourselves. You would be better off putting this question on Crazyswarm's discussion board if you haven't already: https://github.com/USC-ACTLab/crazyswarm/discussions

Let us know if they are able to help you with this.
